### DISCUSSION TITLE
Acknowledgement of data transfered in normal buffering mode

### DISCUSSION THREAD

**Discussion by NewLunarFire** [microsoft/msquic#1350] [Q&A] [answered]
I was reading in the docs that the QUIC_STREAM_EVENT_SEND_COMPLETE event is called immediately after MsQuic copies the data in its internal buffers in the default buffering mode for streams. This allows the application to free its buffers quickly. but when the event is fired the peer has not received or acknowledged the data yet.

Is there a way we can keep track of how much data the peer has acknowledged so far, like get an event when a send is acknowledged? Maybe it is possible using the stats counters within msquic, or maybe adding a QUIC_STREAM_EVENT_DATA_ACKNOWDLEGED? This would be useful for things like keeping track of the progression of a file sent or other large data transfers.

**Comment by nibanks** [comment, marked as answer]
If you want to know how much the peer has acknowledged, then don't use send buffering. If you're willing to deal with the complexity involved you will also get significant perf improvement as well.
